e2e/tests/integration/helpers/mod.rs

```diff
@@ -0,0 +1,228 @@
+use std::io::{self, stderr, stdout, BufRead, Write};
+use std::net::SocketAddr;
+use std::path::{Path, PathBuf};
+use std::process::{Child, Command, ExitStatus, Output, Stdio};
+use std::str;
+use std::thread::sleep;
+use std::time::{Duration, SystemTime};
+
+use colored::*;
+use reqwest::blocking::RequestBuilder;
+
+use lazy_static::lazy_static;
+
+lazy_static! {
+    static ref WORKSPACE_ROOT: PathBuf = {
+        PathBuf::from(env!("CARGO_MANIFEST_DIR"))
+            .parent()
+            .unwrap()
+            .to_path_buf()
+    };
+    static ref DOCKER: PathBuf = which::which("docker").unwrap();
+    static ref DOCKER_COMPOSE: PathBuf = which::which("docker-compose").unwrap();
+    static ref CARGO: PathBuf = which::which("cargo").unwrap();
+    static ref LOCAL_UP: () = {
+        let docker_bake = WORKSPACE_ROOT.join("docker-bake.hcl");
+        let docker_bake_override = WORKSPACE_ROOT.join("e2e/docker-bake-override.hcl");
+        let docker_compose = WORKSPACE_ROOT.join("docker-compose.dev.yml");
```

Comment:
How do the containers shutdown after the test?

Reply:
They don't, they stay up to make the next one (potentially) faster. But I can run a quick clean up `docker compose down` if this sounds like a better idea?

Reply:
I see... hmm, I guess we will have to rebuild and restart them somehow when the API or provisioner code changes

Reply:
Maybe e2e should also not run with every PR but only version branches :man_shrugging:

Reply:
> 

That should not be necessary, as images will then be rebuilt and `docker-compose` will redeploy the services with the new images.

Reply:
> Maybe e2e should also not run with every PR but only version branches

That sounds a bit scary to me for now as our test suite for each service or API element is not super exhaustive, we could end up with frequent merges to main that don't actually work and we would only notice it when we do a release.

Maybe we could break down the test CI action so we have two jobs (and two ticks), that way we can get quicker feedback from unit tests and still rely on the E2E tick before we merge?